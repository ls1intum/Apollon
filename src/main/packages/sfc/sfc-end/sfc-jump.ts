import { UMLElementType } from '../../uml-element-type';
import { ILayer } from '../../../services/layouter/layer';
import { ILayoutable } from '../../../services/layouter/layoutable';
import { SfcContainer } from '../base/sfc-container';
import { Text } from '../../../utils/svg/text';

export class SfcJump extends SfcContainer {
  type = UMLElementType.SfcJump;

  render(canvas: ILayer): ILayoutable[] {
    this.bounds.height = 20;
    this.bounds.width = Text.size(canvas, this.name, { fontWeight: 'bold' }).width + 50;
    return [this];
  }
}
