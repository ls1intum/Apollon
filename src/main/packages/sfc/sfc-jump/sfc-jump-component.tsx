import React from 'react';
import { ThemedPolyline, ThemedRect } from '../../../components/theme/themedComponents';
import { Text } from '../../../components/controls/text/text';
import { SfcJump } from './sfc-jump';

/**
 * Component for rendering a jump element in a sfc.
 * Displays a triangular shape with the jump target name.
 */
export function SfcJumpComponent({ element }: { element: SfcJump }) {
  const sideLength = 20;
  const halfSideLength = sideLength / 2;

  return (
    <g>
      <ThemedRect
        fillColor="#00000000"
        strokeColor="none"
        width={element.bounds.width}
        height={element.bounds.height}
      />
      <ThemedPolyline
        fillColor={element.fillColor}
        strokeColor={element.strokeColor}
        points={`0,0 0,${sideLength} ${sideLength},${halfSideLength} 0,0`}
        strokeWidth="2"
      />
      <Text fill={element.textColor}>{element.name}</Text>
    </g>
  );
}
