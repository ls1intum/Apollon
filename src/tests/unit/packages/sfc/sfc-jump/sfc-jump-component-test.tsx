import { wrappedRender } from '../../../test-utils/render';
import * as React from 'react';
import { SfcJump } from '../../../../../main/packages/sfc/sfc-jump/sfc-jump';
import { SfcJumpComponent } from '../../../../../main/packages/sfc/sfc-jump/sfc-jump-component';

it('render the sfc-jump-component', () => {
  const jump: SfcJump = new SfcJump({ name: 'Jump' });
  const { getByText, baseElement } = wrappedRender(
    <svg>
      <SfcJumpComponent element={jump} />
    </svg>,
  );
  expect(getByText(jump.name)).toBeInTheDocument();
  expect(baseElement).toMatchSnapshot();
});
