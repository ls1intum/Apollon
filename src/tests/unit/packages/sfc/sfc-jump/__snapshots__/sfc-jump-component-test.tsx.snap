// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the sfc-jump-component 1`] = `
<body>
  <div>
    <svg>
      <g>
        <rect
          class="sc-fqkvVR elaLsK"
          fill="#00000000"
          height="100"
          stroke="none"
          width="160"
        />
        <polyline
          class="sc-aXZVg bUvEJV"
          fill="white"
          points="0,0 0,20 20,10 0,0"
          stroke="black"
          stroke-width="2"
        />
        <text
          dominant-baseline="middle"
          font-weight="bold"
          pointer-events="none"
          text-anchor="middle"
          x="50%"
          y="50%"
        >
          Jump
        </text>
      </g>
    </svg>
  </div>
</body>
`;
